File: 0494-target-sum/0494-target-sum.java
class Solution {
    int[] dp;
    public int findTargetSumWays(int[] nums, int target) {
        int total = 0;
        for (int n : nums) {
            total += n;
        }
        target = Math.abs(target);

        if ((target + total) % 2 == 1) {
            return 0;
        }

        int s1 = (target + total) / 2;
        dp = new int[s1 + 1];
        dp[0] = 1;
        find(nums, 0, s1);
        return dp[s1];
    }

    private void find(int[] nums, int index, int target) {
        if (index == nums.length) {
            return;
        }

        for (int n = target; n >= nums[index]; n--) {
            dp[n] = dp[n] + dp[n - nums[index]];
        }

        find(nums, index + 1, target);
    }
}

File: 0494-target-sum/0494-target-sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var findTargetSumWays = function(nums, target) {
    const total = nums.reduce((acc, num) => acc + num, 0);
    target = Math.abs(target);

    if ((target + total) % 2 === 1) {
        return 0;
    }

    const s1 = (target + total) / 2;
    let dp = new Array(s1 + 1).fill(0);
    dp[0] = 1;

    dfs(nums, 0, s1, dp);
    return dp[s1];
};

function dfs(nums, idx, target, dp) {
    if (idx === nums.length) {
        return;
    }

    for (let n = target; n >= nums[idx]; n--) {
        dp[n] = dp[n] + dp[n - nums[idx]];
    }

    dfs(nums, idx + 1, target, dp);
}